torch_geometric/nn/conv/hgt_conv.py

```diff
@@ -143,40 +173,118 @@ def forward(
+        m_rels = [
+            self.m_rel['__'.join(edge_type)] for edge_type in self.edge_types
+        ]
+
+        # List of src node tensors for each node type.
+        q_list = []
+        dst_offset = {}
+        offset = 0
+        for node_type in self.node_types:
+            q_list.append(q_dict[node_type])
+            dst_offset[node_type] = offset
+            offset += q_dict[node_type].size(0)
+        q = torch.cat(q_list)
+
+        # List of edges and edge tensors.
+        edge_index_list = []
+        p_rels = []
+
+        # List of src node tensors for each edge type.
+        k_ins = []
+        v_ins = []
+        src_offset = 0
+        count = 0
+        trans_ptr = [count]
+
+        for edge_type in self.edge_types:
             src_type, _, dst_type = edge_type
-            edge_type = '__'.join(edge_type)
 
-            a_rel = self.a_rel[edge_type]
-            k = (k_dict[src_type].transpose(0, 1) @ a_rel).transpose(1, 0)
-
-            m_rel = self.m_rel[edge_type]
-            v = (v_dict[src_type].transpose(0, 1) @ m_rel).transpose(1, 0)
-
-            # propagate_type: (k: Tensor, q: Tensor, v: Tensor, rel: Tensor)
-            out = self.propagate(edge_index, k=k, q=q_dict[dst_type], v=v,
-                                 rel=self.p_rel[edge_type], size=None)
-            out_dict[dst_type].append(out)
+            # Collect src node tesnors for each edge type.
+            k_src = k_dict[src_type]
+            v_src = v_dict[src_type]
+            k_ins.append(k_src.view(-1, D))
+            v_ins.append(v_src.view(-1, D))
+            for _ in range(H):
+                count += k_src.size(0)
+                trans_ptr.append(count)
+
+            # Collect edges and edge tensors.
+            p_rels.append(self.p_rel['__'.join(edge_type)].view(-1, 1))
+            indices = edge_index_dict[edge_type]
+            # (TODO) Add support for SparseTensor w/o converting?
+            convert = isinstance(indices, SparseTensor)
+            if convert:
+                # Convert to COO
+                dst, src, _ = indices.coo()
+                indices = torch.cat((src.view(1, -1), dst.view(1, -1)))
+            # Add offset to edge indices. Offset needs to be applied as
+            # the src and dst node tensors are concatenated.
+            indices[0] += src_offset
+            indices[1] += dst_offset[dst_type]
+            edge_index_list.append(indices)
+
+            # Update offset.
+            src_offset += k_src.size(0)
+
+        # Concatenate all src node tensors.
+        trans_ptr = torch.tensor(trans_ptr)
+        a_rel, m_rel = torch.cat(a_rels), torch.cat(m_rels)
+        k_out = pyg_lib.ops.segment_matmul(torch.cat(k_ins), trans_ptr,
+                                           a_rel).view(-1, H, D)
+        v_out = pyg_lib.ops.segment_matmul(torch.cat(v_ins), trans_ptr,
+                                           m_rel).view(-1, H, D)
+
+        p = group(p_rels, self.group).view(-1)
```

Comment:
```suggestion
            indices = edge_index_dict[edge_type]
            # (TODO) Add support for SparseTensor w/o converting?
            convert = isinstance(indices, SparseTensor)
            if convert:
                # Convert to COO
                dst, src, _ = indices.coo()
                indices = torch.cat((src.view(1, -1), dst.view(1, -1)))
            num_edges = indices.size(1)
            p_rels.append(self.p_rel['__'.join(edge_type)].view(1, H).expand(num_edges, -1))
            # Add offset to edge indices. Offset needs to be applied as
            # the src and dst node tensors are concatenated.
            indices[0] += src_offset
            indices[1] += dst_offset[dst_type]
            edge_index_list.append(indices)

            # Update offset.
            src_offset += k_src.size(0)

        # Concatenate all src node tensors.
        trans_ptr = torch.tensor(trans_ptr)
        a_rel, m_rel = torch.cat(a_rels), torch.cat(m_rels)
        k_out = pyg_lib.ops.segment_matmul(torch.cat(k_ins), trans_ptr,
                                           a_rel).view(-1, H, D)
        v_out = pyg_lib.ops.segment_matmul(torch.cat(v_ins), trans_ptr,
                                           m_rel).view(-1, H, D)

        # Concatenate all edges and edge tensors.
        p = torch.cat(p_rels, dim = 0)
```
Each edge type should use a different `p`/`rel`. Updating how p is created to reflect that.